eksrollup/lib/k8s.py

```diff
@@ -152,6 +152,131 @@ def taint_node(node_name):
         logger.info("Exception when calling CoreV1Api->patch_node: {}".format(e))
 
 
+def deployment_restart(node_name):
+    """
+    Availability is not guranteed for a single replica deployment.
+    PDB only works with more than 1 replica. Node draining without
+    PDB is synchronous, meaning new pods are started at the same
+    time old pods are being terminated. A deployment restart will
+    ensure there is always one pod available to the service. If
+    enabled, this function will only perform a rolling restart on
+    deployments with single replicas.
+    """
+    # Set threads for the kubectl command
+    threads = []
+
+    ensure_config_loaded()
+
+    k8s_api = client.CoreV1Api()
+
+    apps_k8s_api = client.AppsV1Api()
+
+    # Node field selector used to discover all pods on the node
+    node_field_selector = "spec.nodeName={}".format(node_name)
+
+    if app_config['DRY_RUN'] is True:
+        dry_run="All"
+
+    try:
+        # List all pods on the node
+        pods = k8s_api.list_pod_for_all_namespaces(field_selector=node_field_selector)
+
+        for pod in pods.items:
+
+            pod_name = pod.metadata.name
+            namespace = pod.metadata.namespace
+
+            # Using ownerReferences to verify the pod is owned by a ReplicaSet
+            for pod_owner_reference in pod.metadata.owner_references:
+                if pod_owner_reference.kind == "ReplicaSet":
+                    rs_name = pod_owner_reference.name
+
+                    try:
+                        # Get the ReplicaSet of the pod
+                        rs = apps_k8s_api.read_namespaced_replica_set(rs_name, namespace)
+
+                        # Only rollout restart single replica ReplicaSet
+                        if rs.status.replicas > 1:
+                            logger.debug('pod {} attached to ReplicaSet {} has more than 1 replica. Expecting node draining process to evict pod...'.format(pod_name, rs_name))
+                            continue
+
+                        # Catch if ReplicaSet is 0. Can happen if a deployment has been updated during the run
+                        if rs.status.replicas == 0:
+                            logger.debug('pod {} attached to ReplicaSet {} has 0 replicas...'.format(namespace, rs_name))
```

Comment:
Don't use string formatting in loggers, there's built-in lazy formatting. For example:

```python
logger.debug("pod %s attached to ReplicaSet %s has 0 replicas...", namespace, rs_name)
```

Reply:
awesome, thanks Mike!